cluster-autoscaler/core/scaledown/actuation/update_latency_tracker.go

```diff
@@ -0,0 +1,100 @@
+package actuation
+
+import (
+	"time"
+
+	"k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/taints"
+	"k8s.io/klog/v2"
+)
+
+const sleepDurationWhenPolling = 50 * time.Millisecond
+const watiForTaintingTimeoutDuration = 5 * time.Minute
+
+type NodeTaintStartTime struct {
+	nodeName  string
+	startTime time.Time
+}
+
+type UpdateLatencyTracker struct {
+	startTimestamp     map[string]time.Time
+	finishTimestamp    map[string]time.Time
+	remainingNodeCount int
+	nodeLister         kubernetes.NodeLister
+	startTimeChan      chan NodeTaintStartTime
+	stopChan           chan struct{}
+	resultChan         chan time.Duration
+}
+
+func NewUpdateLatencyTracker(nodeLister kubernetes.NodeLister) *UpdateLatencyTracker {
+	return &UpdateLatencyTracker{
+		startTimestamp:     map[string]time.Time{},
+		finishTimestamp:    map[string]time.Time{},
+		remainingNodeCount: 0,
+		nodeLister:         nodeLister,
+		startTimeChan:      make(chan NodeTaintStartTime),
+		stopChan:           make(chan struct{}),
+		resultChan:         make(chan time.Duration),
+	}
+}
+
+func (u *UpdateLatencyTracker) Start() {
+	for {
+		select {
+		case <-u.stopChan:
+			waitingForTaintingStartTime := time.Now()
```

Comment:
That's a bit too much nesting. Can you extract the logic after `stopChan` is closed to a separate function?

Reply:
Done